src/vnafile.3
.\"
.\" Vector Network Analyzer Library
.\" Copyright © 2020, 2021 D Scott Guthridge <scott_guthridge@rompromity.net>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published
.\" by the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH VNAFILE 3 "NOV 2017" GNU
.nh
.SH NAME
vnafile_alloc, vnafile_check, vnafile_fload, vnafile_free, vnafile_fsave, vnafile_get_dprecision, vnafile_get_file_type, vnafile_get_format, vnafile_get_fprecision, vnafile_load, vnafile_save, vnafile_set_dprecision, vnafile_set_file_type, vnafile_set_format, vnafile_set_fprecision \- Load and Save Network Parameter Data
.\"
.SH SYNOPSIS
.B #include <vnafile.h>
.PP
.BI "vnadata_t *vnafile_alloc(vnafile_error_fn_t *" error_fn ,
.if n \{\
.in +4
.\}
.BI "void *" error_arg );
.if n \{\
.in -4
.\}
.PP
.BI "vnafile_type_t vnafile_get_file_type(const vnafile_t *" vfp );
.PP
.BI "int vnafile_set_file_type(vnafile_t *" vfp ", vnafile_type_t " type );
.PP
.BI "const char *vnafile_get_format(const vnafile_t *" vfp );
.PP
.BI "int vnafile_set_format(vnafile_t *" vfp ", const char *" format );
.PP
.BI "int vnafile_get_fprecision(const vnafile_t *" vfp );
.PP
.BI "int vnafile_set_fprecision(const vnafile_t *" vfp ", int " fprecision );
.PP
.BI "int vnafile_get_dprecision(const vnafile_t *" vfp );
.PP
.BI "int vnafile_set_dprecision(const vnafile_t *" vfp ", int " dprecision );
.PP
.BI "vnafile_t *vnafile_load(const char *" filename ", vnafile_type_t " type ,
.in +4n
.BI "vnafile_error_fn_t *" error_fn ", void " error_arg ", vnadata_t *" vdp );
.in -4n
.PP
.BI "vnafile_t *vnafile_fload(FILE *" fp ", const char *" filename ,
.if n .in +4n
.BI "vnafile_type_t " type ,
.if !n .in +4n
.BI "vnafile_error_fn_t *" error_fn ,
.if n .br
.BI "void " error_arg ", vnadata_t *" vdp );
.in -4n
.PP
.BI "int vnafile_check(vnafile_t *" vfp ", const char *" filename ,
.if n \{\
.in +4
.\}
.BI "const vnadata_t *" vdp );
.if n \{\
.in -4
.\}
.PP
.BI "int vnafile_save(vnafile_t *" vfp ", const char *" filename ,
.if n \{\
.in +4
.\}
.BI "const vnadata_t *" vdp );
.if n \{\
.in -4
.\}
.PP
.BI "int vnafile_fsave(vnafile_t *" vfp ", FILE *" fp ,
.BI "const char *" filename ,
.if n \{\
.in +4
.\}
.BI "const vnadata_t *" vdp );
.if n \{\
.in -4
.\}
.PP
.BI "void vnafile_free(vnafile_t *" vfp );
.PP
Link with \fI-lvna\fP \fI-lm\fP.
.\"
.SH DESCRIPTION
These functions load and save vector network analyzer (VNA) parameter
data in Touchstone and native file formats.
.PP
\fBvnafile_alloc\fP() returns a pointer to a \fBvnafile_t\fP structure
used by the other functions.
The optional \fIerror_fn\fP is a pointer to a function the library
calls with a single line message (without newline) to report errors.
\fIerror_arg\fP is arbitrary user data passed through to the error
function.  Both can be \s-2NULL\s+2.  See \fBvnaerr\fP(3).
Whether an \fIerror_fn\fP is provided or not, the library functions that
can fail set \fBerrno\fP and return -1 or \s-2NULL\s+2 on failure.
.PP
The \fBvnafile_get_file_type\fP() and \fBvnafile_set_file_type\fP()
functions get and set the current file type described by the following
enumeration:
.sp
.in +4n
.nf
.ft CW
typedef enum vnafile_type {
    VNAFILE_AUTO,
    VNAFILE_NATIVE,
    VNAFILE_TOUCHSTONE1,
    VNAFILE_TOUCHSTONE2
} vnafile_type_t;
.ft R
.fi
.in -4n
.sp
The default is \s-2VNAFILE_AUTO\s+2 which causes the load and save
functions to try to automatically intuit the file type based on the
filename extension.
If the file ends with ".s<digit>p", the library assumes Touchstone 1
format; if it ends in ".ts", the library assumes Touchstone 2 format;
if it ends in ".npd" (network parameter data), or anything else,
the library assumes native format.
When loading Touchstone files, the parser internally uses the presence
of the [Version] keyword to determine the version \- either Touchstone
file extension works.
.PP
The \fBvnafile_get_format\fP() and \fBvnafile_set_format\fP() functions
get and set the way VNA parameter data is represented in the file.
The \fIformat\fP parameter is a comma-separated case-insensitive list
of the following specifiers:
.sp
.in +4n
.TS
l l.
S[ri|ma|dB]	scattering parameters
T[ri|ma|dB]	scattering-transfer parameters
Z[ri|ma]	impedance parameters
Y[ri|ma]	admittance parameters
H[ri|ma]	hybrid parameters
G[ri|ma]	inverse-hybrid parameters
A[ri|ma]	ABCD parameters
B[ri|ma]	inverse ABCD parameters
Zin[ri|ma]	impedance looking into each port
PRC	Zin as parallel RC
PRL	Zin as parallel RL
SRC	Zin as series RC
SRL	Zin as series RL
IL	insertion loss
RL	return loss
VSWR	voltage standing wave ratio
.TE
.in -4n
.sp
with the following optional coordinate system modifiers:
.in +4n
.TS
l l.
ri	real, imaginary
ma	magnitude, angle
dB	decibels, angle
.TE
.in -4n
.sp
In the Touchstone file formats, only one specifier can be given and it
must be restricted to one of the s, z, y, h or g variants.
If \fBvnafile_set_format\fP() isn't called, the default format takes
the parameter type from the \fBvnadata_t\fP structure with coordinates
\(lqri\(rq.
.\"
.PP
The \fBvnafile_get_fprecision\fP(), \fBvnafile_set_fprecision\fP(),
\fBvnafile_get_dprecision\fP(), and \fBvnafile_set_dprecision\fP() functions
get and set the numeric precision in decimal digits for frequency and data
values, respectively, when saving to a file.
If not set, \fIfprecision\fP defaults to 7 digits and \fIdprecision\fP
defaults to 6 digits.
.PP
The \fBvnafile_load\fP() function allocates a \fBvnafile_t\fP structure,
loads VNA parameter data from \fIfilename\fP, places the resulting
parameter data into the \fBvnadata_t\fP structure pointed to by \fIvdp\fP
(see \fBvnadata\fP(3)), and stores metadata about the loaded file such
as the file type into the \fBvnafile_t\fP structure.
.PP
The \fBvnafile_loadf\fP() function does the same, but uses the
caller-supplied open file pointer \fIfp\fP.
The \fIfilename\fP parameter to \fBvnafile_loadf\fP() is used only for
error messages and may be given as "-" or other string that doesn't
really refer to a file.
.PP
The \fBvnafile_check\fP() tests if the given combination of options and
parameter data are valid for \fBvnafile_save\fP().
That is, it performs a dry run of save to catch errors before anything
is actually written.
Some examples of things it checks are: that the format string is compatible
with Touchstone formats, that a parameter matrix with dimensions other
than 2x2 isn't being converted to one of the two-port formats
(t, g, h, a or b), and that multiple system impedances aren't given in
Touchstone 1 format.
.PP
The \fBvnafile_save\fP() function saves the VNA parameter data given in
\fIvdp\fP to \fIfilename\fP.
The \fBvnafile_savef\fP() function does the same, except that it uses
the caller-supplied open file pointer \fIfp\fP.
In the later case, \fIvnafile\fP is used for error messages only \-
it doesn't have to refer to an actual file.
.SH "RETURN VALUE"
The \fBvnafile_alloc\fP(), \fBvnafile_load\fP() and \fBvnafile_fload\fP()
function return a pointer to a \fBvnadata_t\fP structure used by the
other functions.
The various get functions return the requested data; they don't return
errors.
The integer-valued functions \fBvnafile_set_file_type\fP(),
\fBvnafile_set_format\fP(), \fBvnafile_set_fprecision\fP(),
\fBvnafile_set_dprecision\fP(), \fBvnfile_check\fP(), \fBvnafile_save\fP()
and \fBvnafile_savef\fP() return zero on success or set \fBerrno\fP and
return -1 on error.
.SH EXAMPLES
.nf
.ft CW
/*
 * Network parameter converter: converts between network parameter
 * types and between Touchstone 1, Touchstone 2 and native file format.
 * The file type is based on filename extension using ".s1p", ".s2p",
 * ".s3p", etc.  for Touchstone 1, ".ts" for Touchstone 2, and ".npd"
 * or other for native format.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <vnafile.h>

static char *progname;

/*
 * usage: usage format
 */
static const char usage[] =
    "%s [-f format] input-file output-file\\n"
    "where format is a comma-separated list of:\\n"
    "  s[ri|ma|dB]  scattering parameters\\n"
    "  t[ri|ma|dB]  scattering-transfer parameters\\n"
    "  z[ri|ma]     impedance parameters\\n"
    "  y[ri|ma]     admittance parameters\\n"
    "  h[ri|ma]     hybrid parameters\\n"
    "  g[ri|ma]     inverse-hybrid parameters\\n"
    "  a[ri|ma]     ABCD parameters\\n"
    "  b[ri|ma]     inverse ABCD parameters\\n"
    "  Zin[ri|ma]   input impedances\\n"
    "  PRC          Zin as parallel RC\\n"
    "  PRL          Zin as parallel RL\\n"
    "  SRC          Zin as series RC\\n"
    "  SRL          Zin as series RL\\n"
    "  IL           insertion loss\\n"
    "  RL           return loss\\n"
    "  VSWR         voltage standing wave ratio\\n"
    "\\n"
    "Coordinates\\n"
    "  ri  real, imaginary\\n"
    "  ma  magnitude, angle\\n"
    "  dB  decibels, angle\\n"
    "\\n"
    "Specifiers are case-insensitive.\\n";

/*
 * error_fn: error printing function for the library
 *   @category: category of error (ignored here)
 *   @message: single line error message without a newline
 *   @error_arg: passed through to the error function (unused here)
 */
static void error_fn(vnaerr_category_t category, const char *message,
        void *error_arg)
{
    (void)fprintf(stderr, "%s: %s\\n", progname, message);
}

/*
 * main
 */
int main(int argc, char **argv)
{
    vnafile_t *vfp;
    vnadata_t *vdp;
    const char *f_opt = NULL;

    if ((char *)NULL == (progname = strrchr(argv[0], '/'))) {
        progname = argv[0];
    } else {
        ++progname;
    }
    for (;;) {
        switch (getopt(argc, argv, "f:")) {
        case -1:
            break;

        case 'f':
            f_opt = optarg;
            continue;

        default:
            (void)fprintf(stderr, usage, progname);
            exit(2);
        }
        break;
    }
    argc -= optind;
    argv += optind;
    if (argc != 2) {
        (void)fprintf(stderr, usage, progname);
        exit(2);
    }
    vdp = vnadata_alloc();
    if ((vfp = vnafile_load(vfp, argv[0], VNAFILE_AUTO,
		    errfn, NULL, vdp)) == NULL) {
        exit(3);
    }
    /* Let vnafile_save determine the file type from the output filename. */
    vnafile_set_file_type(vfp, VNAFILE_AUTO);
    if (f_opt != NULL) {
        if (vnafile_set_format(vfp, f_opt) == -1) {
            exit(4);
        }
    }
    if (vnafile_save(vfp, argv[1], vdp) == -1) {
        exit(5);
    }
    vnadata_free(vdp);
    vnafile_free(vfp);
    exit(0);
}
.ft R
.fi
.\"
.SH "SEE ALSO"
.BR vnacal "(3), " vnacal_new "(3), " vnaconv "(3), " vnadata "(3),"
.BR vnaerr "(3), " vnacal_parameter "(3)"


src/vnafile.c
/*
 * Vector Network Analyzer Library
 * Copyright © 2020, 2021 D Scott Guthridge <scott_guthridge@rompromity.net>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "archdep.h"

#include <ctype.h>
#include <errno.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include "vnafile_internal.h"


/*
 * _vnafile_error report an error
 *   @vfp: pointer to vnafile_t
 *   @category: category of error
 *   @format: printf format string
 */
void _vnafile_error(const vnafile_t *vfp, vnaerr_category_t category,
	const char *format, ...)
{
    va_list ap;

    va_start(ap, format);
    _vnaerr_verror(vfp->vf_error_fn, vfp->vf_error_arg, category, format, ap);
    va_end(ap);
}

/*
 * parse_format: helper function for vnafile_set_format
 *   @vffp: format descriptor
 *   @format: format string
 */
static int parse_format(vnafile_format_t *vffp, const char *format)
{
    const char *cur = format;

    /*
     * Set defaults.
     */
    vffp->vff_parameter = VPT_UNDEF;
    vffp->vff_format = VNAFILE_FORMAT_REAL_IMAG;

    /*
     * Switch on the first letter of the parameter specifier.
     */
    switch (cur[0]) {
    case '\000':
	goto bad_format;

    case 'a':
	++cur;
	vffp->vff_parameter = VPT_A;
	goto parse_coordinates;

    case 'b':
	++cur;
	vffp->vff_parameter = VPT_B;
	goto parse_coordinates;

    case 'd':
	goto parse_coordinates;		/* db */

    case 'g':
	++cur;
	vffp->vff_parameter = VPT_G;
	goto parse_coordinates;

    case 'h':
	++cur;
	vffp->vff_parameter = VPT_H;
	goto parse_coordinates;

    case 'i':
	if (cur[1] == 'l') {
	    cur += 2;
	    vffp->vff_parameter = VPT_S;
	    vffp->vff_format = VNAFILE_FORMAT_IL;
	    break;
	}
	break;

    case 'm':
	goto parse_coordinates;		/* ma */

    case 'p':
	if (strncmp(cur, "prc", 3) == 0) {
	    cur += 3;
	    vffp->vff_parameter = VPT_ZIN;
	    vffp->vff_format = VNAFILE_FORMAT_PRC;
	    break;
	}
	if (strncmp(cur, "prl", 3) == 0) {
	    cur += 3;
	    vffp->vff_parameter = VPT_ZIN;
	    vffp->vff_format = VNAFILE_FORMAT_PRL;
	    break;
	}
	break;

    case 'r':
	if (cur[1] == 'l') {
	    cur += 2;
	    vffp->vff_parameter = VPT_S;
	    vffp->vff_format = VNAFILE_FORMAT_RL;
	    break;
	}
	goto parse_coordinates;		/* ri */

    case 's':
	if (strncmp(cur, "src", 3) == 0) {
	    cur += 3;
	    vffp->vff_parameter = VPT_ZIN;
	    vffp->vff_format = VNAFILE_FORMAT_SRC;
	    break;
	}
	if (strncmp(cur, "srl", 3) == 0) {
	    cur += 3;
	    vffp->vff_parameter = VPT_ZIN;
	    vffp->vff_format = VNAFILE_FORMAT_SRL;
	    break;
	}
	++cur;
	vffp->vff_parameter = VPT_S;
	goto parse_coordinates;

    case 't':
	++cur;
	vffp->vff_parameter = VPT_T;
	goto parse_coordinates;

    case 'v':
	if (strncmp(cur, "vswr", 4) == 0) {
	    cur += 4;
	    vffp->vff_parameter = VPT_S;
	    vffp->vff_format = VNAFILE_FORMAT_VSWR;
	    break;
	}
	break;

    case 'y':
	++cur;
	vffp->vff_parameter = VPT_Y;
	goto parse_coordinates;

    case 'z':
	if (strncmp(cur, "zin", 3) == 0) {
	    cur += 3;
	    vffp->vff_parameter = VPT_ZIN;
	    goto parse_coordinates;
	}
	++cur;
	vffp->vff_parameter = VPT_Z;
	goto parse_coordinates;

    parse_coordinates:
	switch (*cur) {
	case 'd':
	    if (strncmp(cur, "db", 2) == 0) {
		cur += 2;
		vffp->vff_format = VNAFILE_FORMAT_DB_ANGLE;
	    }
	    break;

	case 'm':
	    if (strncmp(cur, "ma", 2) == 0) {
		cur += 2;
		vffp->vff_format = VNAFILE_FORMAT_MAG_ANGLE;
	    }
	    break;

	case 'r':
	    if (strncmp(cur, "ri", 2) == 0) {
		cur += 2;
		vffp->vff_format = VNAFILE_FORMAT_REAL_IMAG;
	    }
	    break;

	default:
	    break;
	}
	break;

    default:
	break;
    }
    if (*cur == '\000') {
	return 0;
    }
bad_format:
    return -1;
}


/*
 * MAX_FORMAT: maximum length of a format specifier
 */
#define MAX_FORMAT	4

/*
 * _vnafile_format_to_name: return the given format descriptor as a string
 *   @vffp: format descriptor
 */
const char *_vnafile_format_to_name(const vnafile_format_t *vffp)
{
    switch (vffp->vff_format) {
    case VNAFILE_FORMAT_PRC:
	return "PRC";

    case VNAFILE_FORMAT_PRL:
	return "PRL";

    case VNAFILE_FORMAT_SRC:
	return "SRC";

    case VNAFILE_FORMAT_SRL:
	return "SRL";

    case VNAFILE_FORMAT_IL:
	return "IL";

    case VNAFILE_FORMAT_RL:
	return "RL";

    case VNAFILE_FORMAT_VSWR:
	return "VSWR";

    default:
	switch (vffp->vff_parameter) {
	case VPT_UNDEF:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "ri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "ma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "dB";
	    default:
		break;
	    }
	    break;

	case VPT_S:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Sri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Sma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "SdB";
	    default:
		break;
	    }
	    break;

	case VPT_T:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Tri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Tma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "TdB";
	    default:
		break;
	    }
	    break;

	case VPT_Z:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Zri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Zma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "ZdB";
	    default:
		break;
	    }
	    break;

	case VPT_Y:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Yri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Yma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "YdB";
	    default:
		break;
	    }
	    break;

	case VPT_H:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Hri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Hma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "HdB";
	    default:
		break;
	    }
	    break;

	case VPT_G:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Gri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Gma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "GdB";
	    default:
		break;
	    }
	    break;

	case VPT_A:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Ari";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Ama";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "AdB";
	    default:
		break;
	    }
	    break;

	case VPT_B:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Bri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Bma";
	    case VNAFILE_FORMAT_DB_ANGLE:
		return "BdB";
	    default:
		break;
	    }
	    break;

	case VPT_ZIN:
	    switch (vffp->vff_format) {
	    case VNAFILE_FORMAT_REAL_IMAG:
		return "Zinri";
	    case VNAFILE_FORMAT_MAG_ANGLE:
		return "Zinma";
	    default:
		break;
	    }
	    break;

	default:
	    break;
	}
    }
    abort();
    /*NOTREACHED*/
}

/*
 * _vnafile_update_format_string: regenerate vfp_format_string
 *   @vfp: pointer to the structure returned from vnafile_alloc
 */
int _vnafile_update_format_string(vnafile_t *vfp)
{
    char *new_string = NULL;
    char *cur;

    free((void *)vfp->vf_format_string);
    vfp->vf_format_string = NULL;
    if ((new_string = malloc(vfp->vf_format_count *
		    (MAX_FORMAT + 1))) == NULL) {
	_vnafile_error(vfp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	return -1;
    }
    cur = new_string;
    for (int i = 0;;) {
	const char *name = _vnafile_format_to_name(&vfp->vf_format_vector[i]);

	(void)strcpy(cur, name);
	cur += strlen(cur);
	if (++i >= vfp->vf_format_count) {
	    break;
	}
	*cur++ = ',';
    }
    *cur = '\000';
    vfp->vf_format_string = new_string;
    return 0;
}

/*
 * vnafile_alloc: allocate the vnafile_t parameter structure
 *   @error_fn:  optional error reporting function
 *   @error_arg: optional opaque argument passed through to the error function
 */
vnafile_t *vnafile_alloc(vnaerr_error_fn_t *error_fn, void *error_arg)
{
    vnafile_t *vfp;

    vfp = (vnafile_t *)malloc(sizeof(vnafile_t));
    if (vfp == NULL) {
	if (error_fn != NULL) {
	    int saved_errno = errno;
	    char message[80];

	    (void)snprintf(message, sizeof(message),
		    "malloc: %s", strerror(errno));
            message[sizeof(message)-1] = '\000';
	    errno = saved_errno;
            (*error_fn)(VNAERR_SYSTEM, message, error_arg);
	    errno = saved_errno;
	}
	return NULL;
    }
    (void)memset((void *)vfp, 0, sizeof(vnafile_t));
    vfp->vf_magic      = VF_MAGIC;
    vfp->vf_error_fn   = error_fn;
    vfp->vf_error_arg  = error_arg;
    vfp->vf_type       = VNAFILE_AUTO;
    vfp->vf_fprecision = 7;
    vfp->vf_dprecision = 6;
    if ((vfp->vf_format_vector = malloc(sizeof(vnafile_format_t))) == NULL) {
	_vnafile_error(vfp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	vnafile_free(vfp);
	return NULL;
    }
    (void)memset((void *)vfp->vf_format_vector, 0, sizeof(vnafile_format_t));
    vfp->vf_format_vector[0].vff_parameter = VPT_UNDEF;
    vfp->vf_format_vector[0].vff_format = VNAFILE_FORMAT_REAL_IMAG;
    vfp->vf_format_count = 1;
    if (_vnafile_update_format_string(vfp) == -1) {
	vnafile_free(vfp);
	return NULL;
    }
    return vfp;
}

/*
 * vnafile_get_file_type: return the file type
 *   @vfp: pointer to the structure returned from vnafile_alloc
 */
vnafile_type_t vnafile_get_file_type(const vnafile_t *vfp)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return (vnafile_type_t)-1;
    }
    return vfp->vf_type;
}

/*
 * vnafile_set_file_type: set the file type
 *   @vfp: pointer to the structure returned from vnafile_alloc
 *   @type: file type
 *
 *   The default type is VNAFILE_AUTO where the library tries to intuit
 *   the type from the filename.
 */
int vnafile_set_file_type(vnafile_t *vfp, vnafile_type_t type)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }
    switch (type) {
    case VNAFILE_AUTO:
    case VNAFILE_NATIVE:
    case VNAFILE_TOUCHSTONE1:
    case VNAFILE_TOUCHSTONE2:
	break;

    default:
	_vnafile_error(vfp, VNAERR_USAGE,
		"vnafile_set_file_type: invalid type");
	return -1;
    }
    vfp->vf_type = type;
    return 0;
}

/*
 * vnafile_get_format: current the format string
 *   @vfp: pointer to the structure returned from vnafile_alloc
 */
const char *vnafile_get_format(const vnafile_t *vfp)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return NULL;
    }
    return vfp->vf_format_string;
}

/*
 * vnafile_set_format: set the format string
 *   @vfp: pointer to the structure returned from vnafile_alloc
 *   @format: a comma-separated case-insensitive list of the following:
 *     [{S,Z,Y,T,H,G,A,B}][{ri,ma,dB}]
 *     {il,rl}
 *     zin[{ri,ma}]
 *     {prc,prl,src,srl}
 *     vswr
 *
 *   If format is NULL or empty string, default to "ri".
 */
int vnafile_set_format(vnafile_t *vfp, const char *format)
{
    vnafile_format_t *vffp_new = NULL;
    size_t length;
    char *format_copy = NULL;
    char *cur;
    int nfields = 1;
    int rc = -1;

    /*
     * Validate pointer.
     */
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }

    /*
     * If format is NULL, default it to "ri".
     */
    if (format == NULL) {
	format = "ri";
    }
    length = strlen(format);

    /*
     * Make a copy of the format string with spaces removed and
     * upper-case converted to lower.  Count comma-separated
     * fields and convert all commas to NUL's.
     */
    if ((format_copy = malloc(length + 1)) == NULL) {
	_vnafile_error(vfp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	return -1;
    }
    cur = format_copy;
    for (const char *cp = format; *cp != '\000'; ++cp) {
	if (*cp > 0x7e) {
	    _vnafile_error(vfp, VNAERR_USAGE, "vnafile_set_format: "
		    "invalid char '\\%02x' in format", *cp);
	    goto out;
	}
	if (isspace(*cp)) {
	    continue;
	}
	if (*cp == ',') {
	    ++nfields;
	    *cur++ = '\000';
	    continue;
	}
	*cur++ = isupper(*cp) ? tolower(*cp) : *cp;
    }
    *cur = '\000';

    /*
     * Allocate a new format vector.
     */
    if ((vffp_new = calloc(nfields, sizeof(vnafile_format_t))) == NULL) {
	_vnafile_error(vfp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	goto out;
    }

    /*
     * Parse the formats and fill in each element of the vector.
     */
    cur = format_copy;
    for (int i = 0;;) {
	if (parse_format(&vffp_new[i], cur) == -1) {
	    _vnafile_error(vfp, VNAERR_USAGE,
		    "invalid format specifier: \"%s\"", cur);
	    goto out;
	}
	if (++i >= nfields) {
	    break;
	}
	cur += strlen(cur) + 1;
    }

    /*
     * Replace the current format vector.
     */
    free((void *)vfp->vf_format_vector);
    vfp->vf_format_vector = vffp_new;
    vffp_new = NULL;
    vfp->vf_format_count = nfields;

    /*
     * Update the format string.
     */
    if (_vnafile_update_format_string(vfp) == -1) {
	goto out;
    }
    rc = 0;

out:
    free((void *)vffp_new);
    free((void *)format_copy);
    return rc;
}

/*
 * _vnafile_set_simple_format: set the format string (1 parameter)
 *   @vfp: pointer to the structure returned from vnafile_alloc
 *   @parameter: parameter type
 *   @format: coordinate system
 */
int _vnafile_set_simple_format(vnafile_t *vfp,
	vnadata_parameter_type_t parameter, vnafile_format_type_t format)
{
    vnafile_format_t *vffp_new = NULL;
    int rc;

    /*
     * Allocate the new format vector, length 1.
     */
    if ((vffp_new = malloc(sizeof(vnafile_format_t))) == NULL) {
	_vnafile_error(vfp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	goto out;
    }
    (void)memset((void *)vffp_new, 0, sizeof(*vffp_new));
    vffp_new->vff_parameter = parameter;
    vffp_new->vff_format = format;

    /*
     * Install the new vector.
     */
    free((void *)vfp->vf_format_vector);
    vfp->vf_format_vector = vffp_new;
    vfp->vf_format_count = 1;

    /*
     * Update the format string.
     */
    if (_vnafile_update_format_string(vfp) == -1) {
	goto out;
    }
    rc = 0;

out:
    return rc;
}

/*
 * vnafile_get_fprecision: get the frequency value precision
 *   @vfp: pointer to the structure returned from vnafile_alloc
 */
int vnafile_get_fprecision(const vnafile_t *vfp)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }
    return vfp->vf_fprecision;
}

/*
 * vnafile_set_fprecision: set the frequency value precision
 *   @vfp: pointer to the structure returned from vnafile_alloc
 *   @precision: precision in decimal places (1..n) or VNAFILE_MAX_PRECISION
 */
int vnafile_set_fprecision(vnafile_t *vfp, int precision)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }
    if (precision < 1) {
	_vnafile_error(vfp, VNAERR_USAGE, "vnafile_set_fprecision: "
		"invalid precision: %d", precision);
	return -1;
    }
    vfp->vf_fprecision = precision;
    return 0;
}

/*
 * vnafile_get_dprecision: set the data value precision
 *   @vfp: pointer to the structure returned from vnafile_alloc
 */
int vnafile_get_dprecision(const vnafile_t *vfp)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }
    return vfp->vf_dprecision;
}

/*
 * vnafile_set_dprecision: set the data value precision for vnafile_save
 *   @vfp: pointer to the structure returned from vnafile_alloc
 *   @precision: precision in decimal places (1..n) or VNAFILE_MAX_PRECISION
 */
int vnafile_set_dprecision(vnafile_t *vfp, int precision)
{
    if (vfp == NULL || vfp->vf_magic != VF_MAGIC) {
	errno = EINVAL;
	return -1;
    }
    if (precision < 1) {
	_vnafile_error(vfp, VNAERR_USAGE, "vnafile_set_fprecision: "
		"invalid precision: %d", precision);
	return -1;
    }
    vfp->vf_dprecision = precision;
    return 0;
}

/*
 * vnafile_free: free the structure obtained from vnafile_alloc
 *   @vfp: pointer to structure to free
 */
void vnafile_free(vnafile_t *vfp)
{
    if (vfp != NULL && vfp->vf_magic == VF_MAGIC) {
	vfp->vf_magic = -1;
	free((void *)vfp->vf_format_vector);
	free((void *)vfp->vf_format_string);
	free((void *)vfp);
    }
}
